### PR TITLE
[diffusion] feat: support LongCat-Image-Edit and LongCat-Image-Edit-Turbo

### PR DESCRIPTION
**Pull request by ITerydh**
## Motivation

SGLang diffusion currently supports LongCat-Image for text-to-image only. This PR adds image-editing (I2I) support for [meituan-longcat/LongCat-Image-Edit](https://huggingface.co/meituan-longcat/LongCat-Image-Edit) and the distilled [meituan-longcat/LongCat-Image-Edit-Turbo](https://huggingface.co/meituan-longcat/LongCat-Image-Edit-Turbo), mirroring the diffusers `LongCatImageEditPipeline` reference implementation.

## Modifications
  - **New `LongCatImageEditPipeline`** (`runtime/pipelines/longcat_image.py`): composed I2I pipeline — input validation → joint VL prompt encoding → image VAE encoding → denoising → decoding.
  - **New `LongCatImageEditPipelineConfig`** (`configs/pipeline_configs/longcat_image.py`), matching the diffusers reference:
    - Output resolution derived from the condition image (~1MP area, ceil to /16), which intentionally differs from the shared /32 helper.
    - Reference image is VAE-encoded (argmax sampling), packed, and concatenated after the noisy latents along the sequence dim; predictions over reference tokens are sliced off.
    - RoPE position ids use modality 1 for noisy tokens and modality 2 for reference tokens, both offset by the full text sequence length (VL image tokens + 512 body).
  - **New `LongCatImageEditTextEncodingStage`** (`stages/model_specific_stages/longcat_image_edit.py`): joint text+image prompt encoding via Qwen2.5-VL, loading the vision tower and masking padded prompt 
  tokens; the negative prompt is encoded against the same image when CFG is enabled. Ulysses SP is supported.
  - **Registry / sampling defaults** (`registry.py`, `configs/sample/longcat_image.py`): Edit (50 steps, CFG 4.5) and Edit-Turbo (8 steps, CFG off); Turbo detector registered first so it wins.
  - **Docs**: compatibility matrix updated with both model ids.
  - **Unit tests**: `test_longcat_image_edit_config.py` (CPU-only config-hook tests) and updated `test_qwen2_5vl_generation.py`.                                                                

  Usage:

  ```bash
CUDA_VISIBLE_DEVICES=0 \
sglang generate \
  --model-path meituan-longcat/LongCat-Image-Edit \
  --image-path demo.jpg \
  --prompt "将图片中人物的衣服改变为蓝色,头发改为红色,保证图像整体细节不变" \
  --num-gpus 1 \
  --num-inference-steps 50 --guidance-scale 4.5 --negative-prompt "" --seed 42 \
  --output-file-path sp1.png --save-output

CUDA_VISIBLE_DEVICES=0,1,2,3,4,5,6,7 \
sglang generate \
  --model-path meituan-longcat/LongCat-Image-Edit \
  --image-path demo.jpg \
  --prompt "将图片中人物的衣服改变为蓝色,头发改为红色,保证图像整体细节不变" \
  --num-gpus 8 --sp-degree 8 --ulysses-degree 8 \
  --num-inference-steps 50 --guidance-scale 4.5 --negative-prompt "" --seed 42 \
  --output-file-path sp8.png --save-output
  ```

## Accuracy Tests

base:
base

result:
result

## Speed Tests and Profiling

    No changes to existing model paths; the new pipeline follows the standard denoising loop. LongCat-Image (T2I) behavior is unchanged. Currently, we are introducing a new model, so there is no need for performance testing. Our main focus is on completing correctness testing.

## Checklist 
- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the 
    speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #32805157172](https://github.com/sgl-project/sglang/actions/runs/32805157172)
Latest PR Test (Extra): :white_check_mark: [Run #32805157023](https://github.com/sgl-project/sglang/actions/runs/32805157023)
Latest PR Test (AMD ROCm 7.2): :hourglass_flowing_sand: [Run #32805157198](https://github.com/sgl-project/sglang/actions/runs/32805157198)

### CODE REVIEW

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/longcat_image_edit.py [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+            self.text_encoder = text_encoder
+
+            pixel_values = vl_inputs["pixel_values"].to(
+                device=device, dtype=self.text_encoder_dtype
+            )
+
+            prompt_embeds = self._encode(
+                prompt, pixel_values, image_grid_thw, prefix_ids, prefix_len, device
+            )
+            batch.prompt_embeds.append(prompt_embeds)
```
Comment (BBuf) [inline python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/longcat_image_edit.py:230]: Does this handle `num_outputs_per_prompt > 1`? `prompt_embeds` stays at batch 1 here, while the image and noise latents use the effective batch size. The DiT will hit a batch mismatch. Could we repeat both positive and negative embeds and add an `n=2` test?

File: python/sglang/multimodal_gen/runtime/models/encoders/qwen2_5vl.py [hunk 2]
```diff
@@ -315,14 +315,13 @@ def forward(
...
-        # causal kernel; its trailing padding is removed during postprocessing.
-        # Cached generation still needs the explicit mask for padded batches.
+        # Honor the prepared mask on the cache-free path too; it is None when
+        # there is nothing to mask, which keeps the native causal kernel.
         attn_output = self.attn(
             query_states,
             key_states,
             value_states,
-            attn_mask=attention_mask if use_cache else None,
+            attn_mask=attention_mask,
```
Comment (BBuf) [inline python/sglang/multimodal_gen/runtime/models/encoders/qwen2_5vl.py]: This also changes the cache-free path for existing LongCat T2I: `longcat_postprocess_text` keeps all 512 body tokens, including padding. Could we add a T2I main/PR or Diffusers parity check and update the PR description? The current claim that T2I behavior is unchanged does not hold.

File: python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/longcat_image_edit.py (hunk 1 above)
Comment (ITerydh) [reply python/sglang/multimodal_gen/runtime/pipelines_core/stages/model_specific_stages/longcat_image_edit.py:230]: Thanks, i was confirmed and fixed. LongCatImagePipelineConfig was missing an expand_conditioning_to_sample_batch override, so on the multi-output path the
  noise/reference latents were built at batch N while prompt_embeds stayed at batch 1, breaking the joint-attention concat. Fixed by mirroring QwenImagePipelineConfig (repeat pos/neg embeds, masks, seq-lens; no-op at n=1), covering both T2I and Edit. Added an n=2 test, and --num-outputs-per-prompt 2   
  now works end-to-end.

File: python/sglang/multimodal_gen/runtime/models/encoders/qwen2_5vl.py (hunk 2 above)
Comment (ITerydh) [reply python/sglang/multimodal_gen/runtime/models/encoders/qwen2_5vl.py]: You're right — the "T2I unchanged" claim doesn't hold, and I've corrected the PR description. Ran the Diffusers parity check you asked for:
                                                                                                                                                                  
  LongCat T2I does feed the padded 512-token body to the DiT (longcat_postprocess_text keeps padding), so masking it changes T2I output. But this is a fix toward parity, not a regression — Diffusers' LongCatImagePipeline._encode_prompt passes attention_mask to the text encoder (pipeline_longcat_image.py:328), so masking the padding is the reference behavior.                                                                                                                          
                                                                                                                                                                  
  Parity check (LongCat-Image, 1024², seed 42, 50 steps, prompt-rewrite off):                                                                                     
  - sglang with mask vs Diffusers ref: mean|Δ| = 13 (chaotic-sampler noise level)
  - sglang without mask vs Diffusers ref: mean|Δ| = 44 — visibly hallucinated background content                                                                  
                                                                                                            
  So the mask brings T2I closer to Diffusers. No separate T2I mechanism is needed since the fix lives in the shared Qwen2.5-VL encoder. Qwen-Image(-Edit) stays unaffected (dynamic padding=True, no pads for a single prompt).

**Review by BBuf — APPROVED** [review]
LGTM

**Comment by BBuf** [comment]
Please fix lint error

**Comment by ITerydh** [comment]
> Please fix lint error

OK

File: python/sglang/multimodal_gen/runtime/models/encoders/qwen2_5vl.py (hunk 2 above)
Comment (BBuf) [reply python/sglang/multimodal_gen/runtime/models/encoders/qwen2_5vl.py]: Could we scope the cache-free mask change to LongCat? Passing a non-None mask switches LocalAttention to the PyTorch SDPA path and changes existing Qwen-Image outputs across multiple backends. I don't think we should update the Qwen-Image GTs in this PR without separate Diffusers parity validation.

Comment (ITerydh) [reply python/sglang/multimodal_gen/runtime/models/encoders/qwen2_5vl.py]: Done — scoped the cache-free mask to LongCat in ec3693e06. Added a honor_cache_free_padding_mask flag that TextEncoderLoader sets only for LongCat (T2I + Edit); it gates both the attn_mask passing and the fallback mask build. Every other pipeline (Qwen-Image included) now keeps the original attn_mask = attention_mask if use_cache else None, so LocalAttention stays on its fast path and no Qwen-Image GTs change in this PR. Verified LongCat sp1/sp8 outputs are byte-identical to before scoping; the unit test asserts both flag states.
